rust/otap-dataflow/crates/engine/src/exporter.rs

```diff
@@ -16,48 +17,68 @@
+//! be thread-safe. If you need to implement an exporter that requires `Send`, you can use the
+//! [`SendEffectHandler`] type. The default effect handler is `!Send` (see
+//! [`NotSendEffectHandler`]).
+//!
+//! # Scalability
+//!
+//! To ensure scalability, the pipeline engine will start multiple instances of the same pipeline
+//! in parallel on different cores, each with its own exporter instance.
 
-use crate::NodeName;
+use crate::config::ExporterConfig;
 use crate::error::Error;
 use crate::message::{ControlMsg, Message};
 use async_trait::async_trait;
 use otap_df_channel::error::RecvError;
 use otap_df_channel::mpsc;
+use std::marker::PhantomData;
+use std::pin::Pin;
 use std::rc::Rc;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::{Instant, Sleep, sleep_until};
 
 /// A trait for egress exporters.
+///
+/// Note: The default effect handler is `!Send` (see [`NotSendEffectHandler`]).
 #[async_trait(?Send)]
-pub trait Exporter {
-    /// The type of messages handled by the exporter.
-    type PData;
-
+pub trait Exporter<PData, EF = NotSendEffectHandler<PData>>
+where
+    EF: EffectHandlerTrait<PData>,
+{
     /// Starts the exporter and begins exporting incoming data.
     ///
     /// The pipeline engine will call this function to start the exporter in a separate task.
     /// Exporters are assigned their own dedicated task at pipeline initialization because their
     /// primary function involves interacting with the external world, and the pipeline has no
     /// prior knowledge of when these interactions will occur.
     ///
-    /// The `Box<Self>` signature indicates that when this method is called, the exporter takes
-    /// exclusive ownership of its instance. This approach is necessary because an exporter cannot
-    /// yield control back to the pipeline engine - it must independently manage its outputs and
-    /// processing timing. The only way the pipeline engine can interact with the exporter after
-    /// starting it is through the control message channel.
+    /// The exporter is taken as `Box<Self>` so it owns itself once `start` is called.
+    /// This lets it move into an independent task, after which the pipeline can only
+    /// reach it through the control-message channel.
+    ///
+    /// Because ownership is now exclusive, the code inside `start` can freely use
+    /// `&mut self` to update internal state without worrying about aliasing or
+    /// borrowing rules at the call-site. That keeps the public API simple (no
+    /// exterior `&mut` references to juggle) while still allowing the exporter to
+    /// mutate itself as much as it needs during its run loop.
     ///
     /// Exporters are expected to process both internal control messages and pipeline data messages,
-    /// prioritizing control messages over data messages.
+    /// prioritizing control messages over data messages. This prioritisation guarantee is ensured
```

Comment:
nit: Use consistent spelling
```suggestion
    /// prioritizing control messages over data messages. This prioritization guarantee is ensured
```